data_science_pipeline/utils/bq.py

```diff
@@ -248,3 +248,66 @@ def run_query_and_save_to_table(  # pylint: disable=too-many-arguments
     if LOGGER.isEnabledFor(logging.DEBUG):
         sample_result = list(islice(result, 3))
         LOGGER.debug("sample_result: %s", sample_result)
+
+
+def does_bq_table_exist(
+        client: Client,
+        project: str,
+        dataset: str,
+        table: str
+):
+    table_id = ".".join([project, dataset, table])
+    try:
+        client.get_table(table_id)
+        LOGGER.info("Table %s already exists.", table_id)
+        return True
+    # pylint disable-broad-except
+    except Exception as e:
+        LOGGER.info("Table not exist. %s", e)
+        return False
+
+
+def create_or_replace_bq_view(
+        query: str,
+        project: str,
+        dataset: str,
+        view_name: str
+):
+    client = get_client(project_id=project)
+    view_id = ".".join([project, dataset, view_name])
+    view = bigquery.Table(view_id)
+    view.view_query = query
+    if not does_bq_table_exist(client, project, dataset, view_name):
+        view = client.create_table(view)
+        LOGGER.info("Created %s: %s", view.table_type, str(view.reference))
+    else:
+        client.delete_table(view_id, not_found_ok=True)
+        LOGGER.info("Deleted view %s", view_id)
+        view = client.create_table(view)
+        LOGGER.info("Recreated %s: %s", view.table_type, str(view.reference))
+
+
+def materialize_given_view(
+        project: str,
+        dataset: str,
+        source_view_name: str,
+        mview_prefix='m'
+):
+    client = get_client(project_id=project)
+    source_view = ".".join([project, dataset, source_view_name])
+    destination_mview_name = "".join([mview_prefix, source_view_name])
+    destination_mview = ".".join([project, dataset, destination_mview_name])
+    LOGGER.info("view %s is been materializing", source_view)
+    job_config = bigquery.QueryJobConfig()
+    sql = '''
+        CREATE OR REPLACE TABLE {destination_mview}
```

Comment:
just a comment in case you are not aware that the the BigQuery API has an explicit option to write the result to a destination table (which is used by https://github.com/elifesciences/bigquery-views-manager/blob/develop/bigquery_views_manager/materialize_views.py and the notebooks that do something similar).

Both options are fine as both are atomic.
(More relevant to the BigQuery Views project, the other option provides also access to the result, e.g. how many rows, which is good for logging)

Reply:
We have logs already from dataframes so that is OK in that case I guess.

Reply:
The logs are a bit different. The dataframe reflects the data that we add to a table. Whereas the logs when materializing views for the BigQuery views project show information like how many rows the materialized view has. e.g.:

```
[2021-09-20 06:08:20,324] {materialize_views.py:44} INFO - materializing view: prod.v_Person_Latest_Contact_Timestamp -> prod.mv_Person_Latest_Contact_Timestamp
[2021-09-20 06:08:25,785] {materialize_views.py:67} INFO - materialized view: prod.v_Person_Latest_Contact_Timestamp, total rows: 157284, 50662021 bytes processed, took: 5.461s
```

But that kind of logging is more valuable to the materialize views DAG because there one can see the same information for each table, making it relatively easy to identify the views that take to long or are very large.